src/cargo/ops/registry.rs

```diff
@@ -55,6 +59,41 @@ pub fn publish(manifest_path: &Path,
     Ok(())
 }
 
+fn check_git_cleanliness(repo: &git2::Repository) -> CargoResult<()> {
+    let mut opts = git2::StatusOptions::new();
+    opts.include_untracked(true).recurse_untracked_dirs(true);
+    let status = try!(repo.statuses(Some(&mut opts)));
+    let files:Vec<String> = status.iter().map(|entry| {
+        let file = entry.index_to_workdir().unwrap()
+             .old_file().path().unwrap().display();
+        format!("{}", file)
+    }).collect();
+    match files.len() {
+        0 => {
+            Ok(())
+        },
+        num @ _ => {
+
+            Err(human(format!("{} uncommited or untacked files \
+        that need to be addressed before publishing. to force the publish command \
+        include --allow-untracked\nproblem files:\n{}", num, files.join("\n"))))
+        }
+    }
+}
+
+fn open_git_repo() -> Result<git2::Repository, git2::Error> {
+    let current_path = Path::new(".");
+    git2::Repository::discover(current_path)
```

Comment:
I am not sure how.  the lack of a known system to check (git)  should return on Ok(()). check_directory_cleanliness logic

```
if git 
  check git
elsif otherSystem
 check other system
else 
  unsupported system return ok. 
```